lib/front_desk.rb

```diff
@@ -0,0 +1,187 @@
+        room_blocks: @room_blocks,
+      )
+      return booking
+    end
+
+    def make_reservation(new_booking_ref, new_number_of_rooms, new_first_room_number, new_check_in, new_check_out)
+      if (@reservations_record.length == 0)
+        booking = reservation_template(new_booking_ref, new_number_of_rooms, new_first_room_number, new_check_in, new_check_out)
+      else
+        for i in (0...@reservations_record.length)
+          old_first_room_no = @reservations_record[i].first_room_number
+          old_last_room_no = @reservations_record[i].first_room_number + @reservations_record[i].number_of_rooms - 1
+          new_last_room_no = new_first_room_number + new_number_of_rooms - 1
+
+          if ((old_first_room_no >= new_first_room_number && old_first_room_no <= new_last_room_no) ||
+              (old_last_room_no >= new_first_room_number && old_last_room_no <= new_last_room_no) ||
+              ((old_first_room_no <= new_first_room_number) && (old_first_room_no <= new_last_room_no) && (old_last_room_no >= new_first_room_number) && (old_last_room_no >= new_last_room_no)))
+            if ((new_check_in >= @reservations_record[i].check_in) && (new_check_in < @reservations_record[i].check_out)) ||
+               ((new_check_out >= @reservations_record[i].check_in) && (new_check_out < @reservations_record[i].check_out))
+              raise ArgumentError, "No BLOCK 4 U. This room is unavailable."
+            else
+              booking = reservation_template(new_booking_ref, new_number_of_rooms, new_first_room_number, new_check_in, new_check_out)
+            end
+          else
+            booking = reservation_template(new_booking_ref, new_number_of_rooms, new_first_room_number, new_check_in, new_check_out)
+          end
+        end
+      end
+      return booking
+    end
+
+    def add_reservation(new_reservation)
+      @reservations_record.push(new_reservation)
+    end
+
+    def get_reservations_by_date(date)
+      reservations_by_date_list = Array.new
+
+      for index in (0...@reservations_record.length)
+        if ((@reservations_record[index].check_in <= date) && (@reservations_record[index].check_out > date))
+          reservations_by_date_list.push(@reservations_record[index])
+        end
+      end
+      return reservations_by_date_list
+    end
+
+    def room_availability(date)
+      reservations_at_date = get_reservations_by_date(date)
+
+      available_rooms = Array.new
+      for room_idx in (0...@all_rooms.length)
+        room_booked = false
+        for res_idx in (0...reservations_at_date.length)
+          for i in (0..@reservations_record[res_idx].room_blocks.length - 1)
+            if (reservations_at_date[res_idx].first_room_number + i == @all_rooms[room_idx])
+              room_booked = true
+              break
+            end
+          end
+        end
+        if (room_booked == false)
+          available_rooms << @all_rooms[room_idx]
+        end
+      end
```

Comment:
This logic isn't clear-- it might be clearer if you iterated through your reservations first, instead of all rooms and then all reservations. However, ultimately, this code isn't readable because of the long variable names and the nested logic. Having a deep "V" nesting (there are 4 levels of nesting) is an indicator that things should be reworked, refactored, and pulled into helper methods.